ros2_graph/graph_generator.py

```diff
@@ -13,425 +13,438 @@
-    @param file_name: File where is stored the ros node info response
-    @return tuples (element, type)
-    """
-    filter_rule = (
-        between_patterns(pattern, file_name)
-        if len(pattern) > 1
-        else after_pattern(pattern[0], file_name)
-    )
-    filter_rule += " | tr -d '[:blank:]'"
-    return tuple(x.split(":") for x in get_clean_lines(filter_rule))
-
-
-def join_name_and_namespace(name: str, namespace: str = None) -> str:
-    """!  Put togheter name and name space on a sigle string
-    @param name
-    @param namespace: ignored for default namespace '/'
-    @return str namespace/name
-    """
-    if namespace is None:
-        return name
-    if len(namespace) > 1:
-        namespace += "/"
-    return namespace + name
+class GraphGenerator:
+    def __init__(self, style_config):
+        rclpy.init()
+        self.dummy = Node("Graph_generator")
+        self.ros_node_info_file = "node_info.txt"
+        self.mainNodes: Dict[int, NodeElement] = dict()
+        self.nodes: Dict[int, NodeElement] = dict()
+        self.topics: Dict[int, NoNodeElement] = dict()
+        self.services: Dict[int, NoNodeElement] = dict()
+        self.actions: Dict[int, NoNodeElement] = dict()
+
+        self.link_strs = style_config["links_str"]
+        self.brackets = style_config["shapes"]
+
+    def newNoNode(
+        self, new_element: NoNodeElement, elemt_dict: Dict[int, NoNodeElement]
+    ) -> NoNodeElement:
+        new_hash = hash(new_element)
+        if new_hash not in elemt_dict:
+            elemt_dict[new_hash] = new_element
+        return elemt_dict[new_hash]
+
+    def newAction(self, data: Tuple[str]) -> NoNodeElement:
+        """! Initialize a NoNodeElement action object form a string tuple (name, namespace, ros_type) or return the already existing action
+        @param action data tuples (name, namespace, ros_type)
+        @return NoNodeElement action object
+        """
+        new_action = NoNodeElement(
+            name=data[0],
+            namespace=data[1],
+            ros_type=data[2],
+            type=ElementType.ACTION,
+            brackets=self.brackets[ElementType.ACTION],
+        )
+        return self.newNoNode(new_action, self.actions)
+
+    def actionsTuple(self, actions_data: Tuple[Tuple[str]]) -> Tuple[NoNodeElement]:
+        """!
+        create a group of actions objects from a tuple of data
+        @param actions_data actions info in (name, namespace, ros_type) format
+        @return a tuple of NoNodeElement action objects
+        """
+        return tuple(self.newAction(data) for data in actions_data)
+
+    def newTopic(self, name: str, namespace: str, ros_type: str) -> NoNodeElement:
+        """! Initialize a NoNodeElement topic object form a string tuple (name, namespace, ros_type) or return the already existing action
+        @param topic data tuples (name, namespace, ros_type)
+        @return NoNodeElement topic object
```

Comment:
why does the method `newAction` receive a tuple and `newTopic` 3 arguments? We should use the same API for similar methods

Reply:
Done